CSI topology e2e tests

verult opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/guide#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/release.md#issue-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/guide/release-notes.md
6. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

/kind bug

**What this PR does / why we need it**: Adds CSI e2e tests for topology, using the GCE PD driver.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #69703

**Special notes for your reviewer**:
* Since the `CSINodeInfo` object has a breaking change, these e2e tests now depend on external-provisioner and external-attacher versions 0.5.x and later. Version 0.4 uses the old CSINodeInfo object. As a result these tests will fail until 0.5 is released. Also, since external-provisioner 0.5 enables Topology by default, there's no need to explicitly set the Topology feature gate in these e2e tests.
* Given time constraints I've deferred fully integrating with the storage e2e testing framework for now.
* Currently these tests fail, but I verified the tests pass without the `CSINodeInfo` change and before rebasing with the new e2e framework. Currently working on verifying them with the e2e framework change.
* Since Topology will be enabled on external-provisioner by default, in CSI provisioning tests it's no longer necessary if the GCE PD driver is multi-zone.

**Does this PR introduce a user-facing change?**: No
<!--  
If no, just write "NONE".
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".
2. 
-->
```release-note
NONE
```

/priority critical-urgent
/sig storage
/assign @msau42 @ddebroy @davidz627 
/cc @mkimuram

verult committed:
CSI topology e2e tests

davidz627 reviewed (commented):
first pass

verult commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -143,6 +145,116 @@ var _ = utils.SIGDescribe("CSI Volumes", func() {
+			driver.CleanupDriver()
+		})
+
+		It("should provision zonal PD with immediate volume binding and mount the volume to a pod", func() {
```
Yep, will remove

verult commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -143,6 +145,116 @@ var _ = utils.SIGDescribe("CSI Volumes", func() {
+			testsuites.TestDynamicProvisioning(t, cs, claim, class)
+		})
+
+		It("should provision zonal PD with immediate volume binding and AllowedTopologies set and mount the volume to a pod", func() {
```
It implicitly verifies by ensuring both positive and negative tests pass. I wanted to avoid inspecting topology details and instead focus on the end user's experience, which is whether or not a volume is available under different topology constraints.

verult commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -304,8 +307,74 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
+	// Create a pod referring to the claim and wait for it to get to running
+	var pod *v1.Pod
+	if t.Unschedulable {
+		pod, err = framework.CreateUnschedulablePod(client, claim.Namespace, t.NodeSelector, []*v1.PersistentVolumeClaim{claim}, true, "")
```
`CreateUnschedulablePod()` creates a pod and waits for it to become Unschedulable, failing if it doesn't within some timeout. It doesn't force the pod to be unschedulable, so it's the former.

I'll rename the field to `ExpectUnschedulable` and clarify the comments too

ddebroy commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -281,10 +284,10 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
 		}
 		command += " || (mount | grep 'on /mnt/test'; false)"
-		runInPodWithVolume(client, claim.Namespace, claim.Name, t.NodeName, command)
+		runInPodWithVolume(client, claim.Namespace, claim.Name, t.NodeName, command, t.NodeSelector, t.Unschedulable)
```
If `t.Unschedulable` is true, there is no need to prepare for IO tests the way this `!t.SkipWriteReadCheck` block does. The two calls to `runInPodWithVolume` also seems redundant if `t.Unschedulable` is true. Will it be a little more intuitive to decouple the testing and verification of `t.Unschedulable` from the `!t.SkipWriteReadCheck` block as `t.Unschedulable == true` implies `t.SkipWriteReadCheck == true`?

verult replied:
It might be easier to do this refactor once `TestBindingWaitForFirstConsumer()` is combined with `TestDynamicProvisioning()`. Would you be OK with that? I can do it as a separate PR.

ddebroy replied:
SGTM

ddebroy reviewed (commented):
Looks good overall with a comment for some refactoring of `TestDynamicProvisioning` if possible. Are multiple PVC tests and regional storage class tests going to come in later PRs?

mkimuram commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -143,6 +145,116 @@ var _ = utils.SIGDescribe("CSI Volumes", func() {
+		driver := drivers.InitGcePDCSIDriver().(drivers.DynamicPVTestDriver) // HACK (verult) eliminate by moving this test to common test suite.
+		BeforeEach(func() {
+			drivers.SetCommonDriverParameters(driver, f, config)
+			driver.SkipUnsupportedTest(testpatterns.TestPattern{}) // HACK gets GCE PD driver to install secrets.
```
Sorry, I wrote the code. I agree that it should be moved to ```CreateDriver()```

davidz627 replied:
could you remove the comment and either fix the functionality or create a TODO+Issue for it

msau42 replied:
This has been fixed in https://github.com/kubernetes/kubernetes/pull/71537, so this can be removed

verult commented:
@ddebroy ah shoot yeah. Tests for Regional PDs aren't specifically necessary to test generic topology functionality, but yeah when GCE PD tests are eventually split off, Regional PDs will need to be tested.

I'll add multi PVC tests in a separate PR

verult commented:
Rebased and addressed comments.

Once external-provisioner 1.0 is cut, the generic provisioning test running against GCE PD driver won't need to skip if the cluster is multizone because the pod will be scheduled in the appropriate zone due to volume node affinity. But for now, I introduced a temporary hack to keep the existing provisioner tests running.

ddebroy commented:
/lgtm

verult commented:
/retest

verult commented:
The last necessary bits should be in now. The tests are failing due to external provisioner missing the latest `CSINodeInfo` changes. Placing a hold for now; this will not go in for 1.13.

/hold

davidz627 commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -143,6 +145,48 @@ var _ = utils.SIGDescribe("CSI Volumes", func() {
 	}
 
+	Context("CSI Topology test using GCE PD driver [Feature: CSINodeInfo]", func() {
+		driver := drivers.InitGcePDCSIDriver().(drivers.DynamicPVTestDriver) // HACK (verult) eliminate by moving this test to common test suite.
```
TODO + Issue or resolve

davidz627 commented on test/e2e/storage/csi_volumes.go:
```diff
@@ -253,6 +297,54 @@ var _ = utils.SIGDescribe("CSI Volumes", func() {
+
+	class := newStorageClass(test, namespace, suffix)
+	if allowedTopologies {
+		topoZone := getRandomCloudZone(cs)
```
`getRandomCloudZone` is a misnomer, should be `getRandomClusterZone`

davidz627 commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -48,6 +48,11 @@ import (
 )
 
+const (
+	GCEPDProvisionerName = "pd.csi.storage.gke.io"
```
nit: clarify these consts are CSI in the name somewhere

davidz627 commented on test/e2e/storage/drivers/csi.go:
```diff
@@ -173,7 +178,12 @@ func (g *gcePDCSIDriver) SkipUnsupportedTest(pattern testpatterns.TestPattern) {
 	config := g.driverInfo.Config
 	framework.SkipUnlessProviderIs("gce", "gke")
-	framework.SkipIfMultizone(cs)
+	if !g.driverInfo.Config.AllAlpha {
```
is `!AllAlpha` the right signal to be skipping on? it seems more like a related signal but not the actual "reason" to skip the test

verult replied:
Whether we skip depends on whether or not topology is enabled, which depends on `AllAlpha`

davidz627 replied:
I guess i'm wondering if there is a "topology enabled" signal we can look for directly, that way we wont have to change this part of the test when toplogy is no longer alpha

msau42 replied:
Can we define some sort of Topology capability and skip tests based off of that?  Similar to plugin capabilities I'm adding here: https://github.com/kubernetes/kubernetes/pull/70756

msau42 replied:
Other drivers that implement topology will probably want this, like ebs.

msau42 replied:
When this moves to beta, we can delete this AllAlpha flag and check entirely, and always run all tests with delayed binding (although this may uncover some problems with test cases that were not written with delayed binding in mind)

verult replied:
Yup I agree the check can be removed after topology is beta, although don't we still want to have tests with immediate binding for full coverage?

Either way once the feature is beta, both immediate and delayed binding will work in both single zone and multizone settings

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -304,8 +307,98 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
+
+	// Wait for ClaimProvisionTimeout (across all PVCs in parallel) and make sure the phase did not become Bound i.e. the Wait errors out
+	By("checking the claims are in pending state")
+	err = framework.WaitForPersistentVolumeClaimsPhase(v1.ClaimBound, client, namespace, claimNames, 2*time.Second, framework.ClaimProvisionShortTimeout, true)
```
nit: name the input variable of `2*timeSecond` with `2*timeSecond /* variable */` or making it an actual variable

davidz627 replied:
this is suspicious, I see `v1.ClaimBound` it looks like we're waiting for that and not for `Pending`?

ddebroy replied:
The verification here (as stated in the comment) is to check that the Claim does not end up in a `Bound` state after the timeout since the mode is `WaitForFirstConsumer`. That is why the `Expect` below is looking for `err` to have occured.

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -304,8 +307,98 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
+	By("creating a pod referring to the claims")
+	// Create a pod referring to the claim and wait for it to get to running
+	var pod *v1.Pod
+	if t.ExpectUnschedulable {
```
these two cases are the same (?) `t.NodeSelector` will be either set or empty right? We can pass it in in both cases. Its unclear to me what difference there is between `CreateUnschedulablePod` and `CreatePod`. Shouldn't they both just be creating a `Pod`, and it is the properties of the pod itself that makes it unschedulable, not any difference in the creation process itself?

ddebroy replied:
`framework.CreatePod` calls `WaitForPodNameRunningInNamespace` and checks the pod is up and running. `framework. CreateUnschedulablePod` calls `WaitForPodNameUnschedulableInNamespace` and checks pod is in an unschedulable state. So that extra wait+verification goes beyond just basic creation of the pod.

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -304,8 +307,98 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
+		framework.WaitForPodToDisappear(client, pod.Namespace, pod.Name, labels.Everything(), framework.Poll, framework.PodDeleteTimeout)
+	}()
+	if t.ExpectUnschedulable {
+		// no claim is provisioned
```
can we check that no claim is provisioned here

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -304,8 +307,98 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
+	if t.ExpectUnschedulable {
+		pod, err = framework.CreateUnschedulablePod(client, namespace, t.NodeSelector, createdClaims, true /* isPrivileged */, "" /* command */)
+	} else {
+		pod, err = framework.CreatePod(client, namespace, nil /* nodeSelector */, createdClaims, true /* isPrivileged */, "" /* command */)
```
do we ever want to check whether this pod actually succeeded in coming up or not?

ddebroy replied:
`CreatePod` already does the waiting: https://github.com/kubernetes/kubernetes/blob/master/test/e2e/framework/pv_util.go#L844

davidz627 commented on test/e2e/storage/testsuites/provisioning.go:
```diff
@@ -304,8 +307,98 @@ func TestDynamicProvisioning(t StorageClassTest, client clientset.Interface, cla
+		Expect(err).NotTo(HaveOccurred())
+		pvs = append(pvs, pv)
+	}
+	Expect(len(pvs)).ToNot(Equal(0))
```
we should actually know exactly how many PV's should be expected, I think it would be `len(createdClaims)`

davidz627 commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -1247,7 +1169,12 @@ func deleteProvisionedVolumesAndDisks(c clientset.Interface, pvs []*v1.Persisten
-	zone, _ := zones.PopAny()
-	return zone
+	if zones.Len() == 0 {
+		return ""
```
this seems like it should be an error case, I don't think any of the callers check for this

davidz627 commented on test/e2e/storage/volume_provisioning.go:
```diff
@@ -1247,7 +1169,12 @@ func deleteProvisionedVolumesAndDisks(c clientset.Interface, pvs []*v1.Persisten
+	}
+
+	zonesList := zones.UnsortedList()
+	return zonesList[rand.Intn(zones.Len())]
```
what was wrong with `PopAny()`?

verult replied:
`PopAny()` is actually deterministic, so it's possible some cases are never tested

davidz627 commented on test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss_alpha.yaml:
```diff
@@ -0,0 +1,58 @@
+      containers:
+        - name: csi-external-provisioner
+          imagePullPolicy: Always
+          image: gcr.io/gke-release/csi-provisioner:v0.4.1-gke.0
```
bump all the versions

verult commented:
Addressed comments; need to update external-provisioner version once it's released

verult commented:
Squashed commits. Unblocking since external provisioner 1.0 has been released.

/hold cancel

verult commented:
/retest

msau42 commented:
/test pull-kubernetes-gce-alpha-features

msau42 commented:
/test pull-kubernetes-e2e-gce-alpha-features

msau42 commented on test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss_alpha.yaml:
```diff
@@ -0,0 +1,55 @@
+      serviceAccountName: csi-controller-sa
+      containers:
+        - name: csi-provisioner
+          image: gcr.io/gke-release/csi-provisioner:v1.0.0-gke.0
```
These should be updated to the latest v1.0.1 sidecars

msau42 commented on test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss_alpha.yaml:
```diff
@@ -0,0 +1,55 @@
+          image: gcr.io/gke-release/csi-provisioner:v1.0.0-gke.0
+          args:
+            - "--v=5"
+            - "--provisioner=pd.csi.storage.gke.io"
```
@davidz627 can we finally get rid of this field? I think v1.0.0+ should include this: https://github.com/kubernetes/kubernetes/pull/71537

davidz627 replied:
wrong link?
but yea I think we should be able to get rid of it, if we finally cherrypicked that change to release-1.0 branch

msau42 commented on test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss_alpha.yaml:
```diff
@@ -0,0 +1,55 @@
+            - name: socket-dir
+              mountPath: /csi
+        - name: gce-pd-driver
+          image: gcr.io/gke-release/gcp-compute-persistent-disk-csi-driver:v0.3.0-gke.0
```
image needs to be updated to v0.3.1-gke.0 after @davidz627 promotes it

verult commented:
Addressed latest comments in a separate commit, will squash if it looks good

msau42 commented:
lgtm

verult commented:
Squashed. Verified tests pass for both single zone and multizone clusters

msau42 commented on test/e2e/testing-manifests/storage-csi/gce-pd/controller_ss.yaml:
```diff
@@ -16,24 +16,23 @@ spec:
       serviceAccountName: csi-controller-sa
       containers:
         - name: csi-provisioner
-          image: gcr.io/gke-release/csi-provisioner:v1.0.0-gke.0
```
Can you also update the daemonset image versions

verult commented:
Addressed comments

msau42 commented:
/lgtm
/approve

As a follow-up, can you look into adding a serial alpha features suite so that the tests will actually run in CI? Or if the alpha features suite is already not running parallely, then modifying the config to allow serial tests
